### PR TITLE
Restrict ADL for `reference_wrapper` and `thread`

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
This PR is unblocking two libcxx tests.
- `std/thread/thread.threads/thread.thread.class/thread.thread.constr/robust_against_adl.pass.cpp`
- `std/utilities/function.objects/refwrap/refwrap.invoke/robust_against_adl.pass.cpp`

Driven-by changes:
- GH-2358 is fixed, but the related libcxx test still fails because it requires `path`'s iterator to satisfy `bidirectional_iterator`. The explanation should be changed.
- Referenced number to the Working Draft in `<type_traits>` and `<thread>` are updated to WG21-N4917 with section numbers added.
- Other ADL-related failures get analyzed.

### CODE REVIEW

**Comment by strega-nil-ms** [comment]
Marking `decision needed`, as we need to discuss the underlying issue #140 this week at the STL maintainers' meeting.

**Comment by StephanTLavavej** [comment]
Removing `decision needed` (as we've decided in #140), and this PR doesn't actually involve qualifying `_Ugly` names.

**Comment by frederick-vs-ja** [comment]
> Removing `decision needed` (as we've decided in #140), and this PR doesn't actually involve qualifying `_Ugly` names.

Oh, I'm sorry for doing too much in one commit.😹 This PR does `_STD`-qualify `_Refwrap_ctor_fun` (only this `_Ugly` name).

File: stl/inc/thread
Comment (strega-nil-ms) [inline stl/inc/thread:314]: Okay this is unrelated but I dislike this behavior immensely and have filed an LWG issue.

**Review by strega-nil-ms — APPROVED** [review]
This looks reasonable to me! Thanks!

File: stl/inc/type_traits
Comment (StephanTLavavej) [inline stl/inc/type_traits:1861]: It is slightly confusing that this comment isn't on the line to which it refers. Since there isn't really space without egregious wrapping, I suggest:
```suggestion
    : true_type {}; // _STD _Refwrap_ctor_fun is qualified: avoid ADL, handle incomplete types
```

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks - I've pushed small changes to fix citations, clarify a comment, and remove unnecessary parens. FYI @strega-nil-ms (as you previously approved) and @CaseyCarter (if you want to double-check the citation that you added in 2019).

**Comment by CaseyCarter** [comment]
I'm not enthusiastic about section numbers in our working draft citations - they're just noise - but it's not worth the effort to pull them out at this point.

**Comment by frederick-vs-ja** [comment]
> I'm not enthusiastic about section numbers in our working draft citations - they're just noise - but it's not worth the effort to pull them out at this point.

I can go with removal of section numbers. Should we make an issue to track this?

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for investigating and fixing these libcxx test failures! :rocket: :heart_eyes_cat: :chart_with_downwards_trend:

**Comment by StephanTLavavej** [comment]
@frederick-vs-ja We already had #182, so I updated that issue to mention that section numbers should be dropped.

**Pull request merged**
